### DISCUSSION TITLE
[A-L] (7 luglio 2023) Esercizio 4

### DISCUSSION THREAD

**Discussion by Frisbon** [sapienzastudentsnetwork/progettazione-di-algoritmi#78] [Esami A-L]
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/155d1e69-421a-4e44-8341-6fc2582cb433)

**Comment by Frisbon** [comment]
Il costo non viene specificato, soluzione naive:
Definisco Kruskal che restituisce un MST dato un grafo (banale)
```python
def es4(G,arco):
    K1 = Kruskal(G)
    if arco not in K1: return "YES"
    K2 = Kruskal(G senza arco)
    
    if len(V(K1)) == len(V(K2)):
        return "YES"
    return "NO"
```
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/842f0ace-c00c-4c14-82d1-4e4a389a35e1)

**Reply by Elia-Belli** [reply]
Il fatto che i sottografi risultanti abbiano un numero diverso di nodi è condizione **sufficiente ma non necessaria** affinché il secondo sottografo non sia un MST, ciò avviene solamente quando l'arco rimosso sconnette $G$ ovvero è un ponte.
Nel caso in cui l'arco rimosso fosse fondamentale per ottenere un albero di copertura di **costo minimo** con l'attuale controllo l'MST sarebbe dato per buono, quindi dobbiamo controllare anche che i pesi degli MST siano uguali.

**Comment by CasuFrost** [comment]
Dato che tutti gli $MST$ di un grafo hanno lo stesso peso, esisterà un $MST$ di $G$ senza $(x,y)$ se e solo se 

$$w(MST(G))=w(MST(G- {(x,y)}))$$

**Comment by Near047** [comment]
es4(G):
  G' = G - {(x,y)}
  Sol = Kruskal(G)
  Sol' = Kruskal(G')
  if Sol' == Sol:             
    return ' DAJE '
 end
  return  'Meno DAJE'


Solo se sono uguali significa che l'arco è inutile.

**Reply by cek753** [reply]
Per un MST va tenuto conto del peso, non è sufficiente dire che hanno lo stesso numero di archi. La soluzione sarebbe quella di modificare kruskal in modo da ritornare il peso, poi fare il confronto tra i due grafi

**Comment by cek753** [comment]
```
def Kruskal(G):
    L=ordino E(G)
    sol={}
    pesoTotale=0
    for e in L Do
        if sol U {e} non contiene cicli:
            sol=sol U {e}
            peso+=w(e)
    return peso
    
Algo(G,(x,y)):
sol1=Kruskal(G)
sol2=Kruskal(G-(x,y))
if sol1==sol2: 
    return 'yes'
return 'no'
```
